File: kotlin-camunda-webcast-app/src/main/java/de/viadee/camunda/demo/mapper/ServiceCallMapper.java
package de.viadee.camunda.demo.mapper;

import de.viadee.camunda.demo.context.ExternalContext;
import de.viadee.camunda.demo.dto.CustomerDTO;
import org.mapstruct.*;

@Mapper(unmappedTargetPolicy = ReportingPolicy.ERROR, componentModel = "spring")
public interface ServiceCallMapper {
    @Mappings({
            @Mapping(source = "street", target = "address.street"),
            @Mapping(source = "zipCode", target = "address.zipCode"),
            @Mapping(source = "city", target = "address.city"),
            @Mapping(source = "country", target = "address.country")
    })
    CustomerDTO map(final ExternalContext externalContext);

    @InheritInverseConfiguration
    @BeanMapping(nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
    @Mappings({
            @Mapping(target = "customerId", ignore = true)
    })
    void map(@MappingTarget final ExternalContext externalContext, final CustomerDTO customer);
}

File: kotlin-camunda-webcast-kotlin/src/main/kotlin/de/viadee/camunda/demo/mapper/ServiceCallMapper.kt
package de.viadee.camunda.demo.mapper

import de.viadee.camunda.demo.context.ExternalContext
import de.viadee.camunda.demo.dto.CustomerDTO
import org.mapstruct.*

@Mapper(unmappedTargetPolicy = ReportingPolicy.ERROR, componentModel = "spring")
interface ServiceCallMapper {
    @Mappings(
        Mapping(source = "street", target = "address.street"),
        Mapping(source = "zipCode", target = "address.zipCode"),
        Mapping(source = "city", target = "address.city"),
        Mapping(source = "country", target = "address.country")
    )
    fun map(externalContext: ExternalContext?): CustomerDTO?

    @InheritInverseConfiguration
    @BeanMapping(nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
    @Mappings(
        Mapping(target = "customerId", ignore = true)
    )
    fun map(@MappingTarget externalContext: ExternalContext?, customer: CustomerDTO?)
}